File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n=nums.size();
        deque<int>dq;
        vector<int>ans;

        //1st Window
        for(int i=0;i<k;i++){
            while(dq.size()>0 && nums[dq.back()]<=nums[i]){
                dq.pop_back();
            }
            dq.push_back(i);
        }
        //remaining windows
        for(int i=k;i<n;i++){
            ans.push_back(nums[dq.front()]);
            int startIdx=i-k+1;
            //jo elements window ke bahar hai usko htao
            while(!dq.empty() && dq.front()<startIdx){
                dq.pop_front();
            }
            while(!dq.empty() && dq.size()>0 && nums[dq.back()]<=nums[i]){
                dq.pop_back();
            }
            dq.push_back(i);
        }
        ans.push_back(nums[dq.front()]);
        return ans;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n=nums.length;

        Deque<Integer>dq=new ArrayDeque<>();
        for(int i=0;i<k;i++){
            while(dq.size()>0 && nums[dq.peekLast()]<=nums[i]){
                dq.removeLast();
            }
            dq.addLast(i);
        }
        ArrayList<Integer>ans= new ArrayList<>();

        for(int i=k;i<n;i++){
            ans.add(nums[dq.peekFirst()]);

            int startIdx=i-k+1;
            while(!dq.isEmpty() && dq.peekFirst()<startIdx){
                dq.removeFirst();
            }

            while(!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]){
                dq.removeLast();
            }

            dq.addLast(i);
        }
        ans.add(nums[dq.peekFirst()]);
        int[] arr=new int[ans.size()];

        for(int i=0;i<ans.size();i++){
            arr[i]=ans.get(i);
        }

        return arr;

    }
}